File: scripts/scenarios/scenarios/44_content_embedding.py
"""Scenario 44: "The Rich Content Creator" — Content Embedding

Luna creates posts with various embed types (images, quote posts,
link cards, record+media) and verifies AppView renders them.

Services: PDS, AppView
"""

# SPDX-FileCopyrightText: 2025-2026 Jack Valinsky
# SPDX-License-Identifier: Unlicense OR CC0-1.0

from __future__ import annotations

import sys
import time
from pathlib import Path

_project_root = str(Path(__file__).resolve().parent.parent.parent.parent)
if _project_root not in sys.path:
    sys.path.insert(0, _project_root)

from scripts.lib.atproto import (
    XrpcClient, get_character, PDS1, ScenarioResult, timed_call,
    SERVICE_URLS, create_account_or_login, now_iso,
)


def run() -> ScenarioResult:
    result = ScenarioResult("Content Embedding")
    result.start()

    pds = XrpcClient(PDS1)
    appview = XrpcClient(SERVICE_URLS["appview"])
    luna = get_character("luna")

    timed_call(result, "PDS health check",
               lambda: pds.wait_for_healthy(timeout=30))
    if result.failed > 0:
        result.finish()
        return result

    # Create account
    session = timed_call(
        result, "Create account for Luna",
        lambda: create_account_or_login(pds, luna),
        detail_fn=lambda s: f"did={s.get('did', '?')}",
    )

    if not session:
        result.finish()
        return result

    luna.did = session["did"]
    luna.access_jwt = session["accessJwt"]

    # Upload an image blob for embedding
    image_blob = timed_call(
        result, "Upload image blob",
        lambda: _upload_test_blob(pds, luna),
        detail_fn=lambda r: f"cid={r.get('cid', '?')}" if r else "failed",
    )

    # 1. Post with image embed
    img_post_rkey = "post-image-" + str(int(time.time()))
    if image_blob:
        img_post = {
            "$type": "app.bsky.feed.post",
            "text": "Check out this image!",
            "createdAt": now_iso(),
            "embed": {
                "$type": "app.bsky.embed.images",
                "images": [{
                    "alt": "A test image",
                    "image": {
                        "$type": "blob",
                        "ref": image_blob.get("ref", {}),
                        "mimeType": "image/png",
                        "size": image_blob.get("size", 0),
                    },
                }],
            },
        }

        img_ref = timed_call(
            result, "Create post with image embed",
            lambda: pds.repositories.create_record(
                luna.did, "app.bsky.feed.post", img_post_rkey,
                img_post, luna.access_jwt,
            ),
            detail_fn=lambda r: f"uri={r.get('uri', '?')}" if r else "failed",
        )
    else:
        result.step_skipped("Create post with image embed", "No blob available")

    # 2. Create a base post for quote embedding
    base_post_rkey = "post-base-" + str(int(time.time()))
    base_post = {
        "$type": "app.bsky.feed.post",
        "text": "Original post to be quoted",
        "createdAt": now_iso(),
    }

    base_ref = timed_call(
        result, "Create base post for quoting",
        lambda: pds.repositories.create_record(
            luna.did, "app.bsky.feed.post", base_post_rkey,
            base_post, luna.access_jwt,
        ),
        detail_fn=lambda r: f"uri={r.get('uri', '?')}" if r else "failed",
    )

    # 3. Post with quote embed
    if base_ref:
        quote_post_rkey = "post-quote-" + str(int(time.time()))
        quote_post = {
            "$type": "app.bsky.feed.post",
            "text": "Quoting this great post!",
            "createdAt": now_iso(),
            "embed": {
                "$type": "app.bsky.embed.record",
                "record": {
                    "uri": base_ref.get("uri", ""),
                    "cid": base_ref.get("cid", ""),
                },
            },
        }

        timed_call(
            result, "Create post with quote embed",
            lambda: pds.repositories.create_record(
                luna.did, "app.bsky.feed.post", quote_post_rkey,
                quote_post, luna.access_jwt,
            ),
        )

    # 4. Post with link card embed (external)
    ext_post_rkey = "post-ext-" + str(int(time.time()))
    ext_post = {
        "$type": "app.bsky.feed.post",
        "text": "Check out this link!",
        "createdAt": now_iso(),
        "embed": {
            "$type": "app.bsky.embed.external",
            "external": {
                "uri": "https://example.com",
                "title": "Example Domain",
                "description": "This domain is for use in illustrative examples",
            },
        },
    }

    timed_call(
        result, "Create post with link card embed",
        lambda: pds.repositories.create_record(
            luna.did, "app.bsky.feed.post", ext_post_rkey,
            ext_post, luna.access_jwt,
        ),
    )

    # Wait for AppView indexing
    time.sleep(2)

    # Verify posts are retrievable via getPosts
    uris_to_check = []
    if base_ref:
        uris_to_check.append(base_ref.get("uri", ""))

    if uris_to_check:
        timed_call(
            result, "Verify posts retrievable via getPosts",
            lambda: appview.feeds.get_posts(
                {"uris": uris_to_check}, luna.access_jwt,
            ),
            detail_fn=lambda r: f"count={len(r.get('posts', []))}" if r else "failed",
        )

    # Verify getPostThread renders embeds
    if base_ref:
        timed_call(
            result, "Get post thread with embeds",
            lambda: appview.feeds.get_post_thread(
                {"uri": base_ref.get("uri", ""), "depth": 2}, luna.access_jwt,
            ),
        )

    result.finish()
    return result


def _upload_test_blob(pds, luna):
    """Upload a small test PNG blob."""
    try:
        # Create a minimal 1x1 PNG
        import base64
        png_data = base64.b64decode(
            "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg=="
        )
        return pds.repositories.upload_blob(
            luna.did, png_data, "image/png", luna.access_jwt,
        )
    except Exception:
        return None

File: scripts/scenarios/scenarios/44_content_embedding.ts
/**
 * @module scenarios/44_content_embedding
 *
 * Scenario: Creates posts with image, quote, and external embeds.
 *
 * Behavior:
 * - Executes the 44 content embedding scenario.
 * - Validates core operations.
 *
 * Expectations:
 * - Scenario completes successfully without errors.
 */

import { getActor, PDS1, SERVICE_URLS } from "../../lib/deno/config.ts";
import { now, ScenarioResult } from "../../lib/deno/runner.ts";
export { ScenarioResult, StepResult, StepStatus } from "../../lib/deno/runner.ts";
export type { ScenarioReport } from "../../lib/deno/runner.ts";
import { XrpcClient, XrpcError } from "../../lib/deno/client.ts";
import { assert } from "../../lib/deno/assertions.ts";
import { timedCall } from "../../lib/deno/runner.ts";

/**
 * Executes the scenario logic.
 * @returns A promise that resolves to the scenario result
 */


const MINIMAL_PNG = new Uint8Array([
  0x89,
  0x50,
  0x4e,
  0x47,
  0x0d,
  0x0a,
  0x1a,
  0x0a,
  0x00,
  0x00,
  0x00,
  0x0d,
  0x49,
  0x48,
  0x44,
  0x52,
  0x00,
  0x00,
  0x00,
  0x01,
  0x00,
  0x00,
  0x00,
  0x01,
  0x08,
  0x02,
  0x00,
  0x00,
  0x00,
  0x90,
  0x77,
  0x53,
  0xde,
  0x00,
  0x00,
  0x00,
  0x0d,
  0x49,
  0x44,
  0x41,
  0x54,
  0x08,
  0xd7,
  0x63,
  0xfc,
  0xff,
  0x9f,
  0xa1,
  0x1e,
  0x00,
  0x07,
  0x82,
  0x02,
  0x3c,
  0x3f,
  0xc8,
  0x48,
  0xef,
  0x00,
  0x00,
  0x00,
  0x00,
  0x49,
  0x45,
  0x4e,
  0x44,
  0xae,
  0x42,
  0x60,
  0x82,
]);

export async function run(): Promise<ScenarioResult> {
  const result = new ScenarioResult("Content Embedding");
  result.start();

  const pds = new XrpcClient(PDS1);
  const appview = new XrpcClient(SERVICE_URLS.appview);
  const luna = getActor("luna");

  await timedCall(result, "PDS health check", async () => {
    await pds.waitForHealthy(30);
  });

  if (result.failed > 0) return result;

  const session = await pds.accounts.createAccount(luna.handle, luna.email, luna.password).catch(
    () => pds.accounts.createSession(luna.handle, luna.password),
  );

  if (!session) {
    result.stepFailed("Setup", "Failed to obtain session");
    result.finish();
    return result;
  }
  luna.did = session.did;
  luna.accessJwt = session.accessJwt;

  const blobResp = await timedCall(result, "Upload image blob", async () => {
    return await pds.blobs.uploadBlob(MINIMAL_PNG, "image/png", luna.accessJwt);
  });

  if (blobResp?.blob) {
    await timedCall(result, "Create post with image embed", async () => {
      return await pds.records.createRecord(luna.did, "app.bsky.feed.post", {
        $type: "app.bsky.feed.post",
        text: "Check out this image!",
        createdAt: now(),
        embed: {
          $type: "app.bsky.embed.images",
          images: [{ alt: "A test image", image: blobResp.blob }],
        },
      }, luna.accessJwt);
    });
  }

  const baseRef = await timedCall(result, "Create base post", async () => {
    return await pds.records.createRecord(luna.did, "app.bsky.feed.post", {
      $type: "app.bsky.feed.post",
      text: "Original post",
      createdAt: now(),
    }, luna.accessJwt);
  });

  if (baseRef) {
    await timedCall(result, "Create quote post", async () => {
      return await pds.records.createRecord(luna.did, "app.bsky.feed.post", {
        $type: "app.bsky.feed.post",
        text: "Quoting!",
        createdAt: now(),
        embed: { $type: "app.bsky.embed.record", record: { uri: baseRef.uri, cid: baseRef.cid } },
      }, luna.accessJwt);
    });
  }

  await timedCall(result, "Create link card post", async () => {
    return await pds.records.createRecord(luna.did, "app.bsky.feed.post", {
      $type: "app.bsky.feed.post",
      text: "Check this out!",
      createdAt: now(),
      embed: {
        $type: "app.bsky.embed.external",
        external: { uri: "https://example.com", title: "Example", description: "Desc" },
      },
    }, luna.accessJwt);
  });

  await new Promise((r) => setTimeout(r, 2000));

  if (baseRef) {
    await timedCall(result, "Verify posts retrievable", async () => {
      return await appview.feed.getPosts([baseRef.uri], luna.accessJwt);
    });

    await timedCall(result, "Get post thread with embeds", async () => {
      return await appview.feed.getPostThread(baseRef.uri, luna.accessJwt);
    });
  }

  result.finish();
  return result;
}

if (import.meta.main) {
  run().then((res) => {
    console.log(res.summary());
    Deno.exit(res.ok ? 0 : 1);
  });
}
